### PR TITLE
Query timeout

### LINKED ISSUE #903: Enable timeout on GRAPH.QUERY runs
enable timeout callbacks that will stop the execution of the query when the client's query reached the ms timeout. 
make usage of `timeout_callback` from `RedisModule_BlockClient` on GRAPH.QUERY

### PR DESCRIPTION
**Pull request by swilly22**
I think we should switch to `RedisModule_CreateTimer` (see [here](https://redis.io/topics/modules-api-ref)) instead of my cron implementation.
Resolves #903

### CODE REVIEW

File: src/commands/cmd_query.c [hunk 1]
```diff
@@ -110,7 +155,9 @@ void Graph_Query(void *args) {
 	QueryCtx_SetResultSet(result_set);
 	if(exec_type == EXECUTION_TYPE_QUERY) {  // query operation
 		ExecutionPlan_PreparePlan(plan);
+		if(timeout != 0 && readonly) Query_SetTimeOut(timeout, plan);
 		result_set = ExecutionPlan_Execute(plan);
+		if(plan->drained) QueryCtx_SetError("Query timed out");
```
Comment (swilly22) [inline src/commands/cmd_query.c:160]: Might want to clear result-set (if possible)

**Comment by filipecosta90** [comment]
@swilly22 
`RedisModuleBlockedClient *RedisModule_BlockClient(RedisModuleCtx *ctx, RedisModuleCmdFunc reply_callback, RedisModuleCmdFunc timeout_callback, void (*free_privdata)(RedisModuleCtx*,void*), long long timeout_ms) already contains the 
`timeout_callback` that will enable to a fp to be called when the timeout is reached in order to send an error to the client ( and cleanup ). Do we need to implement the timeout logic on the module side?

**Comment by jeffreylovitz** [comment]
> `RedisModuleBlockedClient *RedisModule_BlockClient(RedisModuleCtx *ctx, RedisModuleCmdFunc reply_callback, RedisModuleCmdFunc timeout_callback, void (*free_privdata)(RedisModuleCtx*,void*), long long timeout_ms)` already contains the
> `timeout_callback` that will enable to a fp to be called when the timeout is reached in order to send an error to the client ( and cleanup ). Do we need to implement the timeout logic on the module side?

@filipecosta90 We only call `RedisModule_BlockClient` in Lua/MULTI contexts, otherwise we use a NULL `bc` pointer where required.

File: docs/configuration.md [hunk 2]
```diff
@@ -65,3 +65,21 @@ If enabled, RedisGraph will maintain transposed copies of relationship matrices.
...
 $ redis-server --loadmodule ./redisgraph.so MAINTAIN_TRANSPOSED_MATRICES no
 ```
+
+# Query Configurations
+
+Some configurations may be set per query in the form of additional arguments after the query string. All per-query configurations are off by default unless using a language-specific client, which may establish its own defaults.
+
+## Query Timeout
+
+The query flag `--timeout` allows the user to specify the maximum runtime allowed for a query in milliseconds. This configuration can only be set for read queries to avoid leaving the graph in an inconsistent state.
```
Comment (swilly22) [inline docs/configuration.md:75]: remove `--` from `--timeout`

File: src/commands/cmd_query.c [hunk 3]
```diff
@@ -45,11 +47,66 @@ static void _index_operation(RedisModuleCtx *ctx, GraphContext *gc, AST *ast,
...
+				QueryCtx_SetError("Failed to parse query timeout value");
+				return REDISMODULE_ERR;
+			}
+		}
+	}
+	return REDISMODULE_OK;
+}
+
+void QueryTimedOut(void *pdata) {
+	ExecutionPlan *plan = (ExecutionPlan *)pdata;
```
Comment (swilly22) [inline src/commands/cmd_query.c:93]: Add `ASSERT(pdata);`

File: src/commands/cmd_query.c [hunk 4]
```diff
@@ -45,11 +47,66 @@ static void _index_operation(RedisModuleCtx *ctx, GraphContext *gc, AST *ast,
...
+	 * in which case the original query thread had called ExecutionPlan_Free
+	 * decreasing the plan's ref count, but did not free the execution-plan
+	 * it is our responsibility to call ExecutionPlan_Free
+	 *
+	 * Incase execution-plan timedout we'll call ExecutionPlan_Free
+	 * to drop plan's ref count. */
+	ExecutionPlan_Free(plan);
+}
+
+void Query_SetTimeOut(RedisModuleCtx *ctx, uint timeout, ExecutionPlan *plan) {
```
Comment (swilly22) [inline src/commands/cmd_query.c:106]: Remove `RedisModuleCtx *ctx`

File: src/commands/cmd_query.c [hunk 5]
```diff
@@ -82,7 +139,28 @@ void Graph_Query(void *args) {
...
+	int res = _read_flags(command_ctx, &compact, &timeout);
+	if(res == REDISMODULE_ERR) {
+		// Emit error and exit if argument parsing failed.
+		QueryCtx_EmitException();
+		goto cleanup;
+	}
+
+	// Set the query timeout if one was specified.
+	if(timeout != 0) {
+		if(!readonly) {
```
Comment (swilly22) [inline src/commands/cmd_query.c:154]: Combine conditions

File: src/commands/cmd_query.c [hunk 6]
```diff
@@ -111,6 +190,9 @@ void Graph_Query(void *args) {
 	if(exec_type == EXECUTION_TYPE_QUERY) {  // query operation
 		ExecutionPlan_PreparePlan(plan);
 		result_set = ExecutionPlan_Execute(plan);
+
+		if(plan->drained) QueryCtx_SetError("Query timed out"); // Emit error if query timed out.
```
Comment (swilly22) [inline src/commands/cmd_query.c:194]: Change from `plan->drained` to `ExecutionPlan_Drained(plan)`

File: tests/unit/test_cron.cpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	}
+};
+
+TEST_F(CRONTest, CRONTaskExec) {
+	// Add two tasks to CRON
+	// one adds 2 to X
+	// second multiply X by 2
+
+	int *Y = (int*)rm_malloc(sizeof(Y));
+	int *Z = (int*)rm_malloc(sizeof(Z));
```
Comment (swilly22) [inline tests/unit/test_cron.cpp:53]: No need to allocate, use ref

File: src/util/cron.h [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+#pragma once
+
+#include "heap.h"
+#include <sys/types.h>
+
+/* CRON is a task scheduler
+ * a task is defined by:
+ * when it should run; delta in ms from the time it's introduced
+ * a callback to call when it is time to execute the task
+ * and an optional private data passed to the callbacl */
```
Comment (swilly22) [inline src/util/cron.h:16]: `callbacl` `callback`

File: src/util/cron.c [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+	pthread_mutex_t mutex;        // mutex control access to tasks
+	pthread_mutex_t condv_mutex;  // mutex control access to condv
+	pthread_cond_t condv;         // conditional variable
+	pthread_t thread;             // thread running cron main loop
+} CRON;
+
+// single static CRON instance, initialized at CRON_Start
+static CRON *cron = NULL;
+
+static int cmp_timespec(struct timespec a, struct timespec b) {
```
Comment (swilly22) [inline src/util/cron.c:35]: Add comment

File: src/util/cron.c [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+	ASSERT(t);
+	rm_free(t);
+}
+
+static void clear_tasks() {
+	CRON_TASK *task = NULL;
+	while((task = CRON_RemoveTask())) {
+		CRON_FreeTask(task);
+	}
+	heap_free(cron->tasks);
```
Comment (swilly22) [inline src/util/cron.c:118]: Move to `Cron_Stop`

File: src/util/cron.c [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+	pthread_mutex_unlock(&cron->condv_mutex);
+}
+
+// determines if task is due
+static bool CRON_TaskDue(const CRON_TASK *t) {
+	ASSERT(t);
+
+	struct timespec now;
+    clock_gettime(CLOCK_REALTIME, &now);
+	return cmp_timespec(now, t->due) > 0;
```
Comment (swilly22) [inline src/util/cron.c:78]: change to `>=`

**Pull request merged**